pkg/operator/controller/canary/http.go

```diff
@@ -0,0 +1,118 @@
+package canary
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	routev1 "github.com/openshift/api/route/v1"
+
+	"github.com/tcnksm/go-httpstat"
+)
+
+const (
+	echoServerPortAckHeader = "x-request-port"
+)
+
+// probeRouteEndpoint probes the given route's host
+// and returns an error when applicable.
+func probeRouteEndpoint(route *routev1.Route) error {
+	host := route.Spec.Host
+
+	if len(host) == 0 {
+		return fmt.Errorf("route.Spec.Host is empty, cannot test route")
+	}
+
+	// Create HTTP request
+	request, err := http.NewRequest("GET", "http://"+host, nil)
+	if err != nil {
+		return fmt.Errorf("error creating canary HTTP request: %v", err)
+	}
+
+	// Create HTTP result
+	// for request stats tracking.
+	result := &httpstat.Result{}
+
+	// Get request context
+	ctx := httpstat.WithHTTPStat(request.Context(), result)
+	request = request.WithContext(ctx)
+
+	// Send the HTTP request
+	timeout, _ := time.ParseDuration("10s")
+	client := &http.Client{Timeout: timeout}
+	response, err := client.Do(request)
+
+	if err != nil {
+		// Check if err is a DNS error
+		dnsErr := &net.DNSError{}
+		if errors.As(err, &dnsErr) {
+			// Handle DNS error
+			CanaryRouteDNSError.WithLabelValues(host, dnsErr.Server).Inc()
+			return fmt.Errorf("error sending canary HTTP request: DNS error: %v", err)
```

Comment:
log URL? But maybe that would show up in the DNS error string.

Reply:
That will show up in the error string as is